### PR TITLE
enh(API): Added search by host, host group & host category for FindServiceCategories endpoint

### PR DESCRIPTION
**Pull request by dmyios**
`GET: /configuration/services/categories`

Added search parameter `host.id`
Added search parameter `host.name`
Added search parameter `hostgroup.id`
Added search parameter `hostgroup.name`
Added search parameter `hostcategory.id`
Added search parameter `hostcategory.name`

SQL requests:
1. non-admin user, restricted access to hosts, hostgroups and hostcategories (`all_hosts = 0`, `all_hostgroups = 0`)
```
SELECT SQL_CALC_FOUND_ROWS *
FROM (
        SELECT sc.sc_id,
            sc.sc_name,
            sc.sc_description,
            sc.sc_activate
        FROM `centreon`.service_categories sc
            LEFT JOIN `centreon`.service_categories_relation scr ON scr.sc_id = sc.sc_id
            LEFT JOIN `centreon`.host_service_relation hsr ON hsr.service_service_id = scr.service_service_id
            AND hsr.host_host_id IN (
                SELECT arhr.host_host_id
                FROM `centreon`.acl_resources_host_relations arhr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.host ON host.host_id = hsr.host_host_id
        LEFT JOIN `centreon`.hostgroup_relation hr ON hr.host_host_id = host.host_id
            AND hr.hostgroup_hg_id IN (
                SELECT arhgr.hg_hg_id
                FROM `centreon`.acl_resources_hg_relations arhgr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhgr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.hostgroup ON hostgroup.hg_id = hr.hostgroup_hg_id
        LEFT JOIN `centreon`.hostcategories_relation hcr ON hcr.host_host_id = host.host_id
            AND hcr.hostcategories_hc_id IN (
                SELECT arhcr.hc_id
                FROM `centreon`.acl_resources_hc_relations arhcr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhcr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.hostcategories ON hostcategories.hc_id = hcr.hostcategories_hc_id
        INNER JOIN `centreon`.acl_resources_sc_relations arscr ON arscr.sc_id = sc.sc_id
        INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arscr.acl_res_id
        INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
        INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
        WHERE ag.acl_group_id IN (:access_group_ids_1)
            AND sc.level IS NULL
            AND host.host_id IN (:value_1)
        UNION
        SELECT sc.sc_id,
            sc.sc_name,
            sc.sc_description,
            sc.sc_activate
        FROM `centreon`.service_categories sc
            LEFT JOIN `centreon`.service_categories_relation scr ON scr.sc_id = sc.sc_id
            LEFT JOIN `centreon`.service s ON s.service_template_model_stm_id = scr.service_service_id
            LEFT JOIN `centreon`.host_service_relation hsr ON hsr.service_service_id = s.service_id
            AND hsr.host_host_id IN (
                SELECT arhr.host_host_id
                FROM `centreon`.acl_resources_host_relations arhr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.host ON host.host_id = hsr.host_host_id
        LEFT JOIN `centreon`.hostgroup_relation hr ON hr.host_host_id = host.host_id
            AND hr.hostgroup_hg_id IN (
                SELECT arhgr.hg_hg_id
                FROM `centreon`.acl_resources_hg_relations arhgr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhgr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.hostgroup ON hostgroup.hg_id = hr.hostgroup_hg_id
        LEFT JOIN `centreon`.hostcategories_relation hcr ON hcr.host_host_id = host.host_id
            AND hcr.hostcategories_hc_id IN (
                SELECT arhcr.hc_id
                FROM `centreon`.acl_resources_hc_relations arhcr
                    INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arhcr.acl_res_id
                    INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
                    INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
                WHERE ag.acl_group_id IN (:access_group_ids_1)
            )
        LEFT JOIN `centreon`.hostcategories ON hostcategories.hc_id = hcr.hostcategories_hc_id
        INNER JOIN `centreon`.acl_resources_sc_relations arscr ON arscr.sc_id = sc.sc_id
        INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arscr.acl_res_id
        INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
        INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
        WHERE ag.acl_group_id IN (:access_group_ids_1)
            AND sc.level IS NULL
            AND host.host_id IN (:value_2)
    ) AS union_table
```
2. non-admin user access to all hosts, all hostgroups, all hostcategories (`all_hosts = 1`, `all_hostgroups = 1`)
```
SELECT SQL_CALC_FOUND_ROWS *
FROM (
        SELECT sc.sc_id,
            sc.sc_name,
            sc.sc_description,
            sc.sc_activate
        FROM `centreon`.service_categories sc
            LEFT JOIN `centreon`.service_categories_relation scr ON scr.sc_id = sc.sc_id
            LEFT JOIN `centreon`.host_service_relation hsr ON hsr.service_service_id = scr.service_service_id
            LEFT JOIN `centreon`.host ON host.host_id = hsr.host_host_id
            LEFT JOIN `centreon`.hostgroup_relation hr ON hr.host_host_id = host.host_id
            LEFT JOIN `centreon`.hostgroup ON hostgroup.hg_id = hr.hostgroup_hg_id
            LEFT JOIN `centreon`.hostcategories_relation hcr ON hcr.host_host_id = host.host_id
            LEFT JOIN `centreon`.hostcategories ON hostcategories.hc_id = hcr.hostcategories_hc_id
            INNER JOIN `centreon`.acl_resources_sc_relations arscr ON arscr.sc_id = sc.sc_id
            INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arscr.acl_res_id
            INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
            INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
        WHERE ag.acl_group_id IN (:access_group_ids_1)
            AND sc.level IS NULL
            AND host.host_id IN (:value_1)
        UNION
        SELECT sc.sc_id,
            sc.sc_name,
            sc.sc_description,
            sc.sc_activate
        FROM `centreon`.service_categories sc
            LEFT JOIN `centreon`.service_categories_relation scr ON scr.sc_id = sc.sc_id
            LEFT JOIN `centreon`.service s ON s.service_template_model_stm_id = scr.service_service_id
            LEFT JOIN `centreon`.host_service_relation hsr ON hsr.service_service_id = s.service_id
            LEFT JOIN `centreon`.host ON host.host_id = hsr.host_host_id
            LEFT JOIN `centreon`.hostgroup_relation hr ON hr.host_host_id = host.host_id
            LEFT JOIN `centreon`.hostgroup ON hostgroup.hg_id = hr.hostgroup_hg_id
            LEFT JOIN `centreon`.hostcategories_relation hcr ON hcr.host_host_id = host.host_id
            LEFT JOIN `centreon`.hostcategories ON hostcategories.hc_id = hcr.hostcategories_hc_id
            INNER JOIN `centreon`.acl_resources_sc_relations arscr ON arscr.sc_id = sc.sc_id
            INNER JOIN `centreon`.acl_resources res ON res.acl_res_id = arscr.acl_res_id
            INNER JOIN `centreon`.acl_res_group_relations argr ON argr.acl_res_id = res.acl_res_id
            INNER JOIN `centreon`.acl_groups ag ON ag.acl_group_id = argr.acl_group_id
        WHERE ag.acl_group_id IN (:access_group_ids_1)
            AND sc.level IS NULL
            AND host.host_id IN (:value_2)
    ) AS union_table
```

**Fixes** # MON-34392

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by sc979** [comment]
@dmyios , some conflicts require your attention

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php [hunk 1]
```diff
@@ -489,33 +489,182 @@ private function createServiceCategoryFromArray(array $result): ServiceCategory
...
+                        INNER JOIN `:db`.acl_groups ag
+                            ON ag.acl_group_id = argr.acl_group_id
+                        WHERE ag.acl_group_id IN (:access_group_ids)
+                    )
+                    SQL;
+            }
+        }
+
+        $concatenatorA = new SqlConcatenator();
+        $concatenatorA->defineSelect(
```
Comment (jeremyjaouen) [inline centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:552]: Same for ConcatenatorB, can you name the variable more explicitly ? So we could know while reading what is the real difference between both ?

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php [hunk 2]
```diff
@@ -489,33 +489,182 @@ private function createServiceCategoryFromArray(array $result): ServiceCategory
     }
 
     /**
-     * @param SqlConcatenator $concatenator
+     * @param int[] $accessGroupIds
+     *
+     * @return array<SqlConcatenator>
+     */
+    private function findServiceCategoriesRequest(array $accessGroupIds = []): array
```
Comment (jeremyjaouen) [inline centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:496]: maybe use a more convenient name, while reading I expect to return an array of ServiceCategories or a SQL Request  to find the service categories (as a string), but not an array of WHERE condition concatenate

Comment (dmyios) [reply centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:496]: for the naming I followed the same methodolody used for FindServices endpoint where `findServicesRequest` method returns a concatenator

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php [hunk 3]
```diff
@@ -529,6 +754,149 @@ private function retrieveServiceCategories(
...
+                INNER JOIN `:db`.acl_res_group_relations argr
+                    ON argr.acl_res_id = res.acl_res_id
+                INNER JOIN `:db`.acl_groups ag
+                    ON ag.acl_group_id = argr.acl_group_id
+                SQL
+        )->appendWhere(
+            <<<'SQL'
+                WHERE ag.acl_group_id IN (:access_group_ids)
+                SQL
+        );
```
Comment (callapa) [inline centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:888]: No need to use SqlConcatenator.

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php [hunk 4]
```diff
@@ -529,6 +754,149 @@ private function retrieveServiceCategories(
...
```
Comment (callapa) [inline centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:837]: No need to use SqlConcatenator.

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php [hunk 5]
```diff
@@ -529,6 +754,149 @@ private function retrieveServiceCategories(
...
```
Comment (callapa) [inline centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php:788]: No need to use SqlConcatenator.

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php (hunk 3 above)
Comment (dmyios) [reply centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php]: Done

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php (hunk 5 above)
Comment (dmyios) [reply centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php]: Done

File: centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php (hunk 4 above)
Comment (dmyios) [reply centreon/src/Core/ServiceCategory/Infrastructure/Repository/DbReadServiceCategoryRepository.php]: Done

**Review by adr-mo — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**